apps/gimbtonomy/main.cpp

```diff
@@ -47,33 +80,87 @@ void handleSetNeoPixelColor(CANMsg *p_newMsg) {
   }
 }
 
-void processCANMsg(CANMsg *p_newMsg) {
-  switch (p_newMsg->id) {
-    case CANID::SET_NEOPIXEL:
-      // Send an acknowledgement CANMsg back to the Jetson
-      printf("Sending acknowledgement message\r\n");
-      txMsg.clear();
-      txMsg.id = CANID::NEOPIXEL_ACK;
-      txMsg << true;
-      can.write(txMsg);
-      printf("Updating neo pixels\r\n");
-      handleSetNeoPixelColor(p_newMsg);
-      break;
-    default:
-      printf("Recieved unimplemented command for neopixels\r\n");
-      break;
+// Incoming message processor
+void rxCANProcessor() {
+  const int rxPeriod_millisec = 2;
+
+  float pan_pos = 0.0, pan_speed = 0.0, tilt_pos = 0.0, roll_pos = 0.0;
+
+  while (true) {
+    if (can1.read(rxMsg)) {
+      switch (rxMsg.id) {
+        case CANID::SET_GIMBAL_PAN_POS:
+          rxMsg.getPayload(pan_pos);
+          panServoActuator.setMotionData(pan_pos);
+          break;
+
+        case CANID::SET_GIMBAL_PAN_SPEED:
+          rxMsg.getPayload(pan_speed);
+          panServoMotor.servoSetSpeed(pan_speed);
+          break;
+
+        case CANID::SET_GIMBAL_TILT_POS:
+          rxMsg.getPayload(tilt_pos);
+          tiltServo.setPosition(tilt_pos);
+          break;
+
+        case CANID::SET_GIMBAL_ROLL_POS:
+          rxMsg.getPayload(roll_pos);
+          rollServo.setPosition(roll_pos);
+          break;
+
+        case CANID::SET_NEOPIXEL:
+          handleSetNeoPixelColor(&rxMsg);
+          // Hangle can writes in a separate thread
+          event_flags.set(ACK_FLAG);
+          break;
+
+        default:
+          break;
+      }
+    }
+
+    ThisThread::sleep_for(rxPeriod_millisec);
   }
 }
 
-// main() runs in its own thread in the OS
-int main() {
-  printf("Beginning neopixel fw app.\r\n");
-  initCAN();
-  while (1) {
-    if (can.read(rxMsg)) {
-      processCANMsg(&rxMsg);
-      rxMsg.clear();
-    }
+// Outgoing message processor
+void txCANProcessor() {
+  while (true) {
+    // This thread does not sleep, but only executes if signaled by rxCANProcessor to write out a response
+    event_flags.wait_any(ACK_FLAG);
+
+    // Send an acknowledgement CANMsg back to the Jetson
+#ifdef DEBUG
+    printf("Sending neopixel acknowledgement message\r\n");
+#endif
+    txMsg.clear();
+    txMsg.id = CANID::NEOPIXEL_ACK;
+    txMsg << true;
+    can1.write(txMsg);
+#ifdef DEBUG
```

Comment:
👍 I think in the future (not in this PR) we should have a better way to do this kind of tracing so you don't need to use ifdefs everywhere. @wmmc88 mentioned https://github.com/ARMmbed/mbed-trace.